File: Quiz18/Model/GameEngine.swift
import Foundation
import Combine

final class GameEngine: ObservableObject {
    @Published private var gameQuestions: [Question] = []
    @Published private var currentQuestionIndex: Int = 0
    
    private lazy var resourceHandler = ResourceHandler()
    
    public func prepareGame() -> Question {
        guard let questions = resourceHandler.loadResource(ofType: [Question].self, from: Resource.questions) else {
            fatalError("Unable to load questions resource.")
        }
        
        let easy = questions.filter { $0.difficulty == .easy }.shuffled().prefix(5)
        let medium = questions.filter { $0.difficulty == .medium }.shuffled().prefix(5)
        let hard = questions.filter { $0.difficulty == .hard }.shuffled().prefix(5)
        let hardcore = questions.filter { $0.difficulty == .hardcore }.shuffled().prefix(3)
        
        gameQuestions = Array(easy + medium + hard + hardcore)
        currentQuestionIndex = 0
        
        return gameQuestions[currentQuestionIndex]
    }
    
    public func nextQuestion() -> Question? {
        currentQuestionIndex += 1
        
        guard currentQuestionIndex < gameQuestions.count else { return nil }
        
        return gameQuestions[currentQuestionIndex]
    }
    
    public func endGame() -> Game {
        let game = Game(rightAnswersAmount: currentQuestionIndex)
        resourceHandler.saveResource([game], to: Resource.gameHistory)
        
        return game
    }
}

File: android/app/src/main/java/com/quiz18/domain/GameEngine.kt
package com.quiz18.domain

import com.quiz18.data.ResourceHandler

class GameEngine(private val resourceHandler: ResourceHandler) {
    private var gameQuestions: List<Question> = emptyList()
    private var currentQuestionIndex = 0

    fun prepareGame(): Question {
        val questions = resourceHandler.loadQuestions() ?: error("Unable to load questions resource.")
        gameQuestions = questions.filter { it.difficulty == QuestionDifficulty.easy }.shuffled().take(5) +
            questions.filter { it.difficulty == QuestionDifficulty.medium }.shuffled().take(5) +
            questions.filter { it.difficulty == QuestionDifficulty.hard }.shuffled().take(5) +
            questions.filter { it.difficulty == QuestionDifficulty.hardcore }.shuffled().take(3)
        currentQuestionIndex = 0
        return gameQuestions[currentQuestionIndex]
    }

    fun nextQuestion(): Question? {
        currentQuestionIndex += 1
        return gameQuestions.getOrNull(currentQuestionIndex)
    }

    fun endGame(): Game {
        val game = Game(rightAnswersAmount = currentQuestionIndex)
        resourceHandler.saveGame(game)
        return game
    }
}
